Client/Frontend/Home/PanelDataObservers.swift

```diff
@@ -47,29 +47,42 @@ class ActivityStreamDataObserver: DataObserver {
     deinit {
         events.forEach { NotificationCenter.default.removeObserver(self, name: $0, object: nil) }
     }
-    
-    func invalidate(highlights: Bool) {
-        self.delegate?.willInvalidateDataSources()
 
-        let notify = {
-            self.delegate?.didInvalidateDataSources()
+    /*
+     refreshIfNeeded will refresh the underlying caches for both Topsites and Highlights.
+     By default this will only refresh the highlights if the last fetch is older than 15 mins
+     By default this will only refresh topsites if KeyTopSitesCacheIsValid is false
+     */
+    func refreshIfNeeded(forceHighlights highlights: Bool, forceTopSites topSites: Bool) {
+        guard !profile.isShutdown else {
+            return
         }
-        
-        let invalidateTopSites: () -> Success = {
-            self.profile.history.setTopSitesNeedsInvalidation()
-            return self.profile.history.updateTopSitesCacheIfInvalidated() >>> succeed
+
+        // Highlights are cached for 15 mins 200 - 0 > 900 || 200 < 900
+        let invalidateHighlights = highlights ? true : (Timestamp.uptimeInMilliseconds() - lastInvalidation > invalidationTime)
+        lastInvalidation = invalidateHighlights ? Timestamp.uptimeInMilliseconds() : lastInvalidation
```

Comment:
Do this after the invalidation occurred, just before we call `didInvalidateDataSources`.